### PR TITLE
raftstore: report buckets to pd

### PR DESCRIPTION
**Pull request by 5kbpers**
### What is changed and how it works?

Issue Number: Ref https://github.com/tikv/tikv/issues/11718

What's Changed:

```commit-message
Report buckets to pd.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None.
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -4927,6 +4947,38 @@ where
             self.register_reactivate_memory_lock_tick();
         }
     }
+
+    fn on_report_region_buckets_tick(&mut self) {
```
Comment (youjiali1995) [inline components/raftstore/src/store/fsm/peer.rs:4951]: Region counts should be small when dynamic region is enabled, so we don't care about its report frequency?

File: components/raftstore/src/store/worker/read.rs [hunk 2]
```diff
@@ -84,6 +84,7 @@ pub trait ReadExecutor<E: KvEngine> {
         msg: &RaftCmdRequest,
         region: &Arc<metapb::Region>,
         read_index: Option<u64>,
+        bucket_meta: Option<Arc<BucketMeta>>,
```
Comment (youjiali1995) [inline components/raftstore/src/store/worker/read.rs:87]: Reset this file. It's merge fault.

File: components/raftstore/src/store/worker/pd.rs [hunk 3]
```diff
@@ -1601,6 +1633,34 @@ where
...
+                    merge_bucket_stats(
+                        &buckets.meta.keys,
+                        &mut buckets.stats,
+                        &current.meta.keys,
+                        &current.stats,
+                    );
+                    mem::swap(current, &mut buckets);
+                }
+            })
+            .or_insert_with(|| buckets);
```
[the remark this answers is not in the document]
Comment (5kbpers) [reply components/raftstore/src/store/worker/pd.rs]: 😅 Didn't realize this simplification after refactoring.

File: components/raftstore/src/store/fsm/peer.rs (hunk 1 above)
Comment (5kbpers) [reply components/raftstore/src/store/fsm/peer.rs:4951]: Yes, it is supposed to have little impact.

**Review by youjiali1995 — APPROVED** [review]

File: components/pd_client/src/client.rs [hunk 4]
```diff
@@ -933,6 +933,83 @@ impl PdClient for RpcClient {
...
+            let mut inner = client.inner.wl();
+            if let Either::Left(ref mut left) = inner.buckets_sender {
+                debug!("region buckets sender is refreshed");
+                let sender = left.take().expect("expect report region buckets sink");
+                let (tx, rx) = mpsc::unbounded();
+                let pending_buckets = Arc::new(AtomicU64::new(0));
+                inner.buckets_sender = Either::Right(tx);
+                inner.pending_buckets = pending_buckets.clone();
+                let resp = inner.buckets_resp.take().unwrap();
+                inner.client_stub.spawn(async {
```
Comment (tonyxuqqi) [inline components/pd_client/src/client.rs:961]: 疑问: 这里两个spawn()执行顺序和代码一样吗？为什么不能合在一起？

File: components/pd_client/src/client.rs [hunk 5]
```diff
@@ -933,6 +933,83 @@ impl PdClient for RpcClient {
...
+
+        let executor = |client: &Client, req: pdpb::ReportBucketsRequest| {
+            let mut inner = client.inner.wl();
+            if let Either::Left(ref mut left) = inner.buckets_sender {
+                debug!("region buckets sender is refreshed");
+                let sender = left.take().expect("expect report region buckets sink");
+                let (tx, rx) = mpsc::unbounded();
+                let pending_buckets = Arc::new(AtomicU64::new(0));
+                inner.buckets_sender = Either::Right(tx);
+                inner.pending_buckets = pending_buckets.clone();
```
Comment (tonyxuqqi) [inline components/pd_client/src/client.rs:959]: 疑问，这里为什么叫pending_buckets, 从逻辑上看更像是已经发送的buckets count

File: components/pd_client/src/client.rs (hunk 4 above)
Comment (youjiali1995) [reply components/pd_client/src/client.rs:961]: They are concurrent.

File: components/pd_client/src/client.rs (hunk 5 above)
Comment (youjiali1995) [reply components/pd_client/src/client.rs:959]: It counts the buckets in the mpsc queue.

**Comment by tonyxuqqi** [comment]
LGTM

File: components/raftstore/src/store/worker/split_check.rs [hunk 6]
```diff
@@ -377,6 +383,17 @@ where
                 policy,
             } => self.check_split_and_bucket(&region, auto_split, policy),
             Task::ChangeConfig(c) => self.change_cfg(c),
+            Task::ApproximateBuckets(region) => {
+                if self.coprocessor.cfg.enable_region_bucket {
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/split_check.rs:387]: If it's not enabled, the task should not be sent in the first place.

**Review by BusyJay — COMMENTED** [review]
So many duplicated code.

File: components/raftstore/src/store/worker/split_check.rs (hunk 6 above)
Comment (5kbpers) [reply components/raftstore/src/store/worker/split_check.rs:387]: Added a check before sending tasks.

**Review by BusyJay — APPROVED** [review]

**Pull request merged**